github/GithubRetry.py

```diff
@@ -0,0 +1,229 @@
+#                                                                              #
+# Copyright 2022 Enrico Minack <github@enrico.minack.dev>                      #
+#                                                                              #
+# This file is part of PyGithub.                                               #
+# http://pygithub.readthedocs.io/                                              #
+#                                                                              #
+# PyGithub is free software: you can redistribute it and/or modify it under    #
+# the terms of the GNU Lesser General Public License as published by the Free  #
+# Software Foundation, either version 3 of the License, or (at your option)    #
+# any later version.                                                           #
+#                                                                              #
+# PyGithub is distributed in the hope that it will be useful, but WITHOUT ANY  #
+# WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS    #
+# FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more #
+# details.                                                                     #
+#                                                                              #
+# You should have received a copy of the GNU Lesser General Public License     #
+# along with PyGithub. If not, see <http://www.gnu.org/licenses/>.             #
+#                                                                              #
+################################################################################
+
+import json
+import logging
+from datetime import datetime, timezone
+from logging import Logger
+from typing import Optional
+
+from requests import Response
+from requests.models import CaseInsensitiveDict
+from requests.utils import get_encoding_from_headers
+from urllib3 import HTTPResponse, Retry
+from urllib3.exceptions import MaxRetryError
+
+from github.GithubException import GithubException
+from github.Requester import Requester
+
+DEFAULT_SECONDARY_RATE_WAIT: int = 60
+
+
+class GithubRetry(Retry):
+    """
+    A Github-specific implementation of `urllib3.Retry`
+
+    This retries 403 responses if they are retry-able. Github requests are retry-able when
+    the response provides a `"Retry-After"` header, or the content indicates a rate limit error.
+
+    By default, response codes 403, and 500 up to 599 are retried. This can be configured
+    via the `status_forcelist` argument.
+
+    By default, all methods defined in `Retry.DEFAULT_ALLOWED_METHODS` are retried, plus GET and POST.
+    This can be configured via the `allowed_methods` argument.
+    """
+
+    __logger: Optional[Logger] = None
+
+    # used to mock datetime, mock.patch("github.GithubRetry.date") does not work as this
+    # references the class, not the module (due to re-exporting in github/__init__.py)
+    __datetime = datetime
```

Comment:
I see you mention mocking here, but the changes don't seem to actually do any mocking. Am I missing something?

Reply:
The mocking is done here:
https://github.com/PyGithub/PyGithub/pull/2387/files#diff-223d619ef9fda22ba4275f05a88a10b6403d6735aa31b8985be6062a910bd288R124-R138